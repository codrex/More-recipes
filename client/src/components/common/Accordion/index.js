export { default as Accordion } from './Accordion';
export { default as AccordionHeader } from './AccordionHeader';
export { default as AccordionBody } from './AccordionBody';
