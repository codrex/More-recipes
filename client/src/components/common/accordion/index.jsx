export { default as Accordion } from './accordion';
export { default as AccordionHeader } from './accordionHeader';
export { default as AccordionBody } from './accordionBody';
