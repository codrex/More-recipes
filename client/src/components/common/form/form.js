import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import './form.scss';

export const Input = (props) => (
  <div className="form-group">
    <input
      type={props.type || 'text'}
      className={classnames('form-control text-input', props.className)}
      id={props.id}
      aria-describedby={props.id}
      placeholder={props.placeholder}
      onChange={props.handleChange}
      value={props.value}
    />
  </div>
);

Input.propTypes = {
  type: PropTypes.string,
  id: PropTypes.string,
  placeholder: PropTypes.string,
  value: PropTypes.string,
  className: PropTypes.string,
  handleChange: PropTypes.func,
};

export const Form = (props) => (
  <form
    className={classnames('col-12 no-margin', props.className)}
    action={props.action}
    method={props.method}
    onSubmit={e => {
      e.preventDefault();
      props.onSubmit();
    }}
  >
    {props.children}
    <button
      type="submit"
      className="btn btn-secondary btn-lg text-white text-uppercase w-100"
    >{props.submitBtnText}
    </button>
  </form>
);

Form.propTypes = {
  children: PropTypes.arrayOf(PropTypes.element).isRequired,
  action: PropTypes.string,
  method: PropTypes.string,
  className: PropTypes.string,
  submitBtnText: PropTypes.string.isRequired,
  onSubmit: PropTypes.func,
};
