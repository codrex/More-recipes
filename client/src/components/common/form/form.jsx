import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

const Form = props => (
  <form
    className={classnames('col-12', props.className)}
    action={props.action}
    method={props.method}
    onSubmit={e => {
      e.preventDefault();
      props.onSubmit();
    }}
  >
    {props.children}
    {props.submitBtnText && <button
      type="submit"
      className={classnames('btn text-uppercase w-100',
      props.secondary && 'btn-secondary',
      props.primary && 'btn-primary',
      props.primaryInverse && 'btn-primary-inverse',
      props.secondaryInverse && 'btn-secondary-inverse',
      props.lg && 'btn-lg',
      props.disabled && 'disable'
      )}
      disabled={props.disabled}
    >{props.disabled && 'loading...' || props.submitBtnText}
    </button>}
  </form>
);

Form.defaultProps = {
  lg: true
};

Form.propTypes = {
  children: PropTypes.any,
  action: PropTypes.string,
  method: PropTypes.string,
  className: PropTypes.string,
  submitBtnText: PropTypes.string,
  onSubmit: PropTypes.func,
  primary: PropTypes.bool,
  secondary: PropTypes.bool,
  lg: PropTypes.bool,
  disabled: PropTypes.bool,
  primaryInverse: PropTypes.bool,
  secondaryInverse: PropTypes.bool,
};

export default Form;
