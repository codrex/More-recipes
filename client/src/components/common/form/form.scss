@import '../../../scss/_var.scss';

.text-input{
  min-height: 2.5em;
  font-size: 2em;
  transition: box-shadow .3s ease-in;
}
.help-text{
  font-size: 1.2rem;
  color: $danger-color;
  font-weight: 200;
}
