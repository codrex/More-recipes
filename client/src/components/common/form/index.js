export { default as Form } from './form';
export { default as Input } from './input';
export { default as PureInput } from './pureInput';
export { default as Select } from './select';
export { default as Textarea } from './textarea';
export { default as PureTextarea } from './pureTextarea';
