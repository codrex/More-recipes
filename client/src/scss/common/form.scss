
.text-input {
  min-height: 2.5em;
  font-size: 1.4em;
  transition: box-shadow 0.3s ease-in;
}
.help-text {
  font-size: 1.2rem;
  color: $danger-color;
  font-weight: 200;
  padding-left: 0.5em;
  padding-top: 0.25em;
  font-weight: 600;
  &::first-letter {
    text-transform: uppercase;
  }
}
.disable{
  box-shadow: none !important;
}
