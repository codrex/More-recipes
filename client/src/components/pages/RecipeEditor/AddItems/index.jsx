import React from 'react';
import PropTypes from 'prop-types';
import ItemsList from './ItemsList';
import { Input, Textarea, Form } from '../../../common/FormElements';
import { validateItem } from '../../../../utils/validator';
import ItemField from './ItemField';
import { RECIPE_ADDED } from '../../../../constants';

/**
 * React component to add recipe Items
 */
class AddItems extends React.Component {
  /**
   * constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      items: props.items || [],
      showInput: true,
      validationError: props.externalError
    };
  }

  /**
   * @param {Object} nextProps
   * @return {boolean} boolean value
   */
  componentWillReceiveProps = (nextProps) => {
    if (this.state.validationError !== nextProps.externalError) {
      this.setState({
        validationError: nextProps.externalError
      });
    }
    this.setState({
      items: nextProps.items
    });
  }

  /**
   * @param {object} nextProps
   * @return {boolean} boolean value
   */
  shouldComponentUpdate = (nextProps) => {
    if (nextProps.message === RECIPE_ADDED) {
      return false;
    }
    return true;
  }

  /**
   * add item to list
   * @param {String} value
   * @return {undefined} undefined
   */
  addItem = (value) => {
    const { directions, ingredients } = this.props;
    const { items } = this.state;
    const item = value[this.props.name].toLowerCase();

    const itemsUpdate = (ingredients && [item].concat(items)) ||
    (directions && items.concat(item));

    this.props.sendItemsToStore(itemsUpdate);
    this.props.initialize();
  }

  /**
   * remove item from list
   * @param {number} index
   * @return {undefined} undefined
   */
  deleteFromList = (index) => {
    const items = [].concat(this.state.items);
    items.splice(index, 1);
    this.props.sendItemsToStore(items);
  }

  /**
   * edit item
   * @param {string} value
   * @param {number} index
   * @return {undefined} undefined
   */
  editItems = (value, index) => {
    const items = [].concat(this.state.items);
    items[index] = value;
    this.props.sendItemsToStore(items);
  }

  /**
   * validate item
   * @return {Array} array of strings
   * @param {string} value
   */
  validate = (value) => {
    const { name } = this.props;
    const lowerCaseValue = value && value.toLowerCase();
    if (this.state.items.includes(lowerCaseValue)) {
      return [`${name} is already on the list`];
    }
    return validateItem(lowerCaseValue, name);
  }

  /**
   * render form
   * @return {React} react component
   */
  renderForm = () => {
    const {
      handleSubmit,
      name,
      placeholder,
      ingredients,
      directions,
      clearValidationError
    } = this.props;
    const Component = (ingredients && Input) || (directions && Textarea);
    if (this.state.showInput) {
      return (
        <Form
          onSubmit={handleSubmit(this.addItem)}
          submitBtnText={`add ${name}`}
          secondary
          lg={false}
          id={`${name}-submit-btn`}
        >
          <ItemField
            Component={Component}
            name={name}
            placeholder={placeholder}
            validate={this.validate}
            clearValidationError={clearValidationError}
            validationError={this.state.validationError}
          />
        </Form>
      );
    } return null;
  }

  /**
   * render list
   * @return {React} react component
   */
  renderList = () => {
    const {
      name,
      Component,
      ingredients,
      directions
    } = this.props;
    return (
      <ItemsList
        items={this.state.items}
        deleteItem={this.deleteFromList}
        editItem={this.editItems}
        componentType={name}
        Component={Component}
        name={name}
        directions={directions}
        ingredients={ingredients}
        uniquenessChecker={this.validate}
      />
    );
  }

  /**
   * render
   * @return {React} react component
   */
  render = () => {
    const { name } = this.props;
    const { items } = this.state;
    return (
      <div className="col-12 items ">
        <div className="items-header">
          <h4 className="items-header-text">
            {`${name}s ${' '}${items.length}`}
          </h4>
          {this.renderForm()}
        </div>
        <div className="col-12 items-list-wrapper">
          {this.renderList()}
        </div>
      </div>
    );
  }
}

AddItems.defaultProps = {
  ingredients: false,
  directions: false,
  externalError: [],
  Component: null
};

AddItems.propTypes = {
  clearValidationError: PropTypes.func.isRequired,
  directions: PropTypes.bool,
  externalError: PropTypes.oneOfType(
    [
      PropTypes.string,
      PropTypes.array, PropTypes.bool
    ]
  ),
  handleSubmit: PropTypes.func.isRequired,
  ingredients: PropTypes.bool,
  initialize: PropTypes.func.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape).isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  sendItemsToStore: PropTypes.func.isRequired,
  Component: PropTypes.oneOfType(
    [
      PropTypes.node,
      PropTypes.element,
      PropTypes.func
    ]
  )
};

export default AddItems;
