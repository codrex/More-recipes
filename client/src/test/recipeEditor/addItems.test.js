import React from 'react';
import toJson from 'enzyme-to-json';
import AddItems from '../../components/pages/recipeEditor/addItems';
import { PureInput } from '../../components/common/form';

const props = {
  handleSubmit: jest.fn(),
  placeholder: '',
  name: 'ingredient',
  initialize: jest.fn(),
  clearValidationError: jest.fn(),
  externalError: [],
  items: [],
  directions: true,
  sendItemsToStore: jest.fn(),
  ingredients: false,
  // Component: PureInput
};

describe('Add items component ', () => {
  test('render as expected when directions is passed as props', () => {
    const component = shallow(
      <AddItems {...props} />
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
  test('render as expected when form is submitted', () => {
    const component = shallow(
      <AddItems {...props} ingredients />
    );
    component.find('Form').simulate('submit');
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
  test('render as expected when there is external validation error', () => {
    const component = shallow(
      <AddItems {...props} ingredients />
    );
    component.setProps({ externalError: ['error'] });
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
  test('render as expected when name field has focus', () => {
    const component = shallow(
      <AddItems {...props} ingredients />
    );
    component.find('Field').simulate('focus');
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
