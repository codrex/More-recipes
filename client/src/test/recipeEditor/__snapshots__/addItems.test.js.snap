// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add items component  render as expected when directions is passed as props 1`] = `
<div
  className="col-12 items "
>
  <div
    className="items-header"
  >
    <h4
      className="items-header-text"
    >
      ingredients  0
    </h4>
    <Form
      action=""
      className=""
      disabled={false}
      lg={false}
      method=""
      onSubmit={[Function]}
      primary={false}
      primaryInverse={false}
      secondary={true}
      secondaryInverse={false}
      style={Object {}}
      submitBtnText="add ingredient"
    >
      <Field
        className="no-margin add-item-input"
        component={[Function]}
        externalError={
          Object {
            "error": Array [],
            "touched": true,
          }
        }
        fgClassName="d-flex flex-column no-margin"
        id="ingredient"
        name="ingredient"
        onFocus={[Function]}
        placeholder=""
        type="text"
        validate={[Function]}
      />
    </Form>
  </div>
  <div
    className="col-12 items-list-wrapper"
  >
    <List
      Component={null}
      componentType="ingredient"
      deleteItem={[Function]}
      directions={true}
      editItem={[Function]}
      ingredients={false}
      items={Array []}
      name="ingredient"
      uniquenessChecker={[Function]}
    />
  </div>
</div>
`;

exports[`Add items component  render as expected when form is submitted 1`] = `
<div
  className="col-12 items "
>
  <div
    className="items-header"
  >
    <h4
      className="items-header-text"
    >
      ingredients  0
    </h4>
    <Form
      action=""
      className=""
      disabled={false}
      lg={false}
      method=""
      onSubmit={[Function]}
      primary={false}
      primaryInverse={false}
      secondary={true}
      secondaryInverse={false}
      style={Object {}}
      submitBtnText="add ingredient"
    >
      <Field
        className="no-margin add-item-input"
        component={[Function]}
        externalError={
          Object {
            "error": Array [],
            "touched": true,
          }
        }
        fgClassName="d-flex flex-column no-margin"
        id="ingredient"
        name="ingredient"
        onFocus={[Function]}
        placeholder=""
        type="text"
        validate={[Function]}
      />
    </Form>
  </div>
  <div
    className="col-12 items-list-wrapper"
  >
    <List
      Component={null}
      componentType="ingredient"
      deleteItem={[Function]}
      directions={true}
      editItem={[Function]}
      ingredients={true}
      items={Array []}
      name="ingredient"
      uniquenessChecker={[Function]}
    />
  </div>
</div>
`;

exports[`Add items component  render as expected when name field has focus 1`] = `
<div
  className="col-12 items "
>
  <div
    className="items-header"
  >
    <h4
      className="items-header-text"
    >
      ingredients  0
    </h4>
    <Form
      action=""
      className=""
      disabled={false}
      lg={false}
      method=""
      onSubmit={[Function]}
      primary={false}
      primaryInverse={false}
      secondary={true}
      secondaryInverse={false}
      style={Object {}}
      submitBtnText="add ingredient"
    >
      <Field
        className="no-margin add-item-input"
        component={[Function]}
        externalError={
          Object {
            "error": Array [],
            "touched": true,
          }
        }
        fgClassName="d-flex flex-column no-margin"
        id="ingredient"
        name="ingredient"
        onFocus={[Function]}
        placeholder=""
        type="text"
        validate={[Function]}
      />
    </Form>
  </div>
  <div
    className="col-12 items-list-wrapper"
  >
    <List
      Component={null}
      componentType="ingredient"
      deleteItem={[Function]}
      directions={true}
      editItem={[Function]}
      ingredients={true}
      items={Array []}
      name="ingredient"
      uniquenessChecker={[Function]}
    />
  </div>
</div>
`;

exports[`Add items component  render as expected when there is external validation error 1`] = `
<div
  className="col-12 items "
>
  <div
    className="items-header"
  >
    <h4
      className="items-header-text"
    >
      ingredients  0
    </h4>
    <Form
      action=""
      className=""
      disabled={false}
      lg={false}
      method=""
      onSubmit={[Function]}
      primary={false}
      primaryInverse={false}
      secondary={true}
      secondaryInverse={false}
      style={Object {}}
      submitBtnText="add ingredient"
    >
      <Field
        className="no-margin add-item-input"
        component={[Function]}
        externalError={
          Object {
            "error": Array [
              "error",
            ],
            "touched": true,
          }
        }
        fgClassName="d-flex flex-column no-margin"
        id="ingredient"
        name="ingredient"
        onFocus={[Function]}
        placeholder=""
        type="text"
        validate={[Function]}
      />
    </Form>
  </div>
  <div
    className="col-12 items-list-wrapper"
  >
    <List
      Component={null}
      componentType="ingredient"
      deleteItem={[Function]}
      directions={true}
      editItem={[Function]}
      ingredients={true}
      items={Array []}
      name="ingredient"
      uniquenessChecker={[Function]}
    />
  </div>
</div>
`;
